### PR TITLE
Remove the 'hold' kwarg from codebase

### PR DESCRIPTION
**Pull request by dstansby**
Another step in to the bright new post python 2.7 world, this time removing the `hold` kwarg. As far as I can see I've got rid of all the `_hold` instances in the code base. Lets see what happens with the tests...

### CODE REVIEW

File: lib/matplotlib/pyplot.py [hunk 1]
```diff
@@ -2510,203 +2428,83 @@ def spy(Z, precision=0, marker=None, markersize=None, aspect='equal', **kwargs):
...
-        ax._hold = washold
+    ret = ax.acorr(x, data=data, **kwargs)
 
     return ret
 
 # Autogenerated by boilerplate.py.  Do not edit as changes will be lost.
 @_autogen_docstring(Axes.angle_spectrum)
 def angle_spectrum(x, Fs=None, Fc=None, window=None, pad_to=None, sides=None,
-                   hold=None, data=None, **kwargs):
+                   **kwargs):
```
Comment (dstansby) [inline lib/matplotlib/pyplot.py]: Note to future me: the data kwarg shouldn't have gone here, this needs fixing. (or if anyone fancies force pushing a fix to my branch)

**Comment by tacaswell** [comment]
The +/- on this PR is amazing!

**Comment by QuLogic** [comment]
I think this is somewhat covered by #9173, though it's a bit old and has some conflicts.

**Comment by efiring** [comment]
There is overlap with #9173, but I think it would be reasonable to merge this once the tests pass, and then revisit the approach in #9173.
It looks like the test problem is a single missed `_hold` in tripcolor.

**Comment by anntzer** [comment]
I'll need to revisit #9173 anyways, so no worries.

**Comment by tacaswell** [comment]
Just needs API changes docs!

File: tools/boilerplate.py [hunk 2]
```diff
@@ -49,19 +49,7 @@
...
-    if hold is not None:
-        %(ax)s._hold = hold
-        from matplotlib.cbook import mplDeprecation
-        warnings.warn("The 'hold' keyword argument is deprecated since 2.0.",
-                      mplDeprecation)
-    try:
-        %(ret)s = %(ax)s.%(func)s(%(call)s)
-    finally:
-        %(ax)s._hold = %(washold)s
+    %(ret)s = %(ax)s.%(func)s(%(call)s)
```
Comment (anntzer) [inline tools/boilerplate.py]: Even simpler would be to just do `return gca().func(...)`, which avoids the need for temporary variables and thus allows you to get rid of the "gensym-like facility" mentioned lower in this script.

Comment (tacaswell) [reply tools/boilerplate.py]: :+1: the only reason we needed the local reference to the return value was so we could reset the hold.

**Comment by anntzer** [comment]
Still a few fixes applicable but let's get this in first.

**Pull request merged**